Customers API: Fix exact match filters

psealock opened the pull request:
Fixes https://github.com/woocommerce/woocommerce-admin/issues/1617

Customers Report advanced filters by `country`, `email`, `name`, and `usersname` now can include/exclude by exact match (in other words, no fuzzy search).

### Screenshots

![Screen Shot 2019-03-20 at 8 49 58 AM](https://user-images.githubusercontent.com/1922453/54637308-36e20f00-4aed-11e9-92ab-87f3669ed5d2.png)

### Detailed test instructions:

1. Customers Report > Advanced Filters.
2. Filter by country, email, name, and username using included and excluded.
3. Ensure the filters are working correctly.

psealock committed:
Customers API: fix exact match params

timmyc reviewed (approved):
Code is looking good and testing well for me, advanced filters are ✨ on the customers report!

jeffstieler commented on includes/api/class-wc-admin-rest-reports-customers-controller.php:
```diff
@@ -350,18 +355,43 @@ public function get_collection_params() {
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+		$params['email_includes']                   = array(
+			'description'       => __( 'Limit response to objects including to an emails.', 'woocommerce-admin' ),
```
Nitpick: grammar. _(same for excludes)_

> including to an emails

jeffstieler commented on includes/api/class-wc-admin-rest-reports-customers-stats-controller.php:
```diff
@@ -211,18 +216,43 @@ public function get_collection_params() {
+			'validate_callback' => 'rest_validate_request_arg',
+		);
+		$params['email_includes']                   = array(
+			'description'       => __( 'Limit response to objects including to an emails.', 'woocommerce-admin' ),
```
Same grammar nitpick.

jeffstieler commented on includes/data-stores/class-wc-admin-reports-customers-data-store.php:
```diff
@@ -216,17 +216,25 @@ protected function get_sql_query_params( $query_args ) {
-				); // WPCS: unprepared SQL ok.
+			if ( ! empty( $query_args[ $exact_match_param . '_includes' ] ) ) {
+				$exact_match_arguments = $query_args[ $exact_match_param . '_includes' ];
+				$table_column          = 'country' === $exact_match_param ? $exact_match_param : 'customer_id';
```
A comment explaining this ternary would be helpful.

jeffstieler commented on includes/data-stores/class-wc-admin-reports-customers-data-store.php:
```diff
@@ -216,17 +216,25 @@ protected function get_sql_query_params( $query_args ) {
+			if ( ! empty( $query_args[ $exact_match_param . '_includes' ] ) ) {
+				$exact_match_arguments = $query_args[ $exact_match_param . '_includes' ];
+				$table_column          = 'country' === $exact_match_param ? $exact_match_param : 'customer_id';
+				$included              = implode( "','", explode( ',', $exact_match_arguments ) );
```
These need to get SQL escaped to protect against injection. _(same with excludes)_

jeffstieler reviewed (commented):
This tests well - but the SQL injection needs to get taken care of. Small grammar nitpicks as well.

psealock commented:
Thanks for the review @jeffstieler!

I fixed the grammar errors, added a comment about the ternary, and used `esc_sql` before adding the query arguments to the sql statements.

jeffstieler commented on includes/data-stores/class-wc-admin-reports-customers-data-store.php:
```diff
@@ -216,17 +216,27 @@ protected function get_sql_query_params( $query_args ) {
+			}
+
+			if ( ! empty( $query_args[ $exact_match_param . '_excludes' ] ) ) {
+				$exact_match_arguments = esc_sql( $query_args[ $exact_match_param . '_excludes' ] );
```
`esc_sql()` needs to get called for each item in this array as we're building the `WHERE` clause.

jeffstieler commented on includes/data-stores/class-wc-admin-reports-customers-data-store.php:
```diff
@@ -250,7 +260,7 @@ protected function get_sql_query_params( $query_args ) {
 		// Allow a list of customer IDs to be specified.
 		if ( ! empty( $query_args['customers'] ) ) {
-			$included_customers = implode( ',', $query_args['customers'] );
+			$included_customers = implode( ',', esc_sql( $query_args['customers'] ) );
```
`esc_sql()` is meant for string values that will be in quotes in the SQL statement - we should probably `intval()` each customer ID instead.

jeffstieler commented on includes/data-stores/class-wc-admin-reports-customers-data-store.php:
```diff
@@ -216,17 +216,27 @@ protected function get_sql_query_params( $query_args ) {
-					$query_args[ $exact_match_param ]
-				); // WPCS: unprepared SQL ok.
+			if ( ! empty( $query_args[ $exact_match_param . '_includes' ] ) ) {
+				$exact_match_arguments = esc_sql( $query_args[ $exact_match_param . '_includes' ] );
```
`esc_sql()` needs to get called for each item in this array as we're building the `WHERE` clause.

psealock commented:
Thanks for the second round of feedback @jeffstieler.

I ensured `esc_sql` is applied to each element in the array of exact match arguments and used `intval` on each of the `$query_args['customers']` values.

jeffstieler reviewed (approved):
Code looks good and tests well. 🚢it!
